File: 模拟/规律/对角线遍历/对角线遍历.py
# https://leetcode.cn/problems/diagonal-traverse/

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        i, j, s, n, m, ans = 0, 0, 1, len(mat), len(mat[0]), []
        while True:
            if j >= m:
                i += 2
                j -= 1
                s = -s
            elif i >= n:
                i -= 1
                j += 2
                s = -s
            elif j <  0:
                j += 1
                s = -s
            elif i <  0:
                i += 1
                s = -s
            ans.append(mat[i][j])
            if i==n-1 and j==m-1:
                break
            i -= s
            j += s
        return ans

File: 模拟/规律/对角线遍历/对角线遍历.cpp
// https://leetcode.cn/problems/diagonal-traverse/

class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        vector<int> ans;
        int i = 0, j = 0, s = 1, n = mat.size(), m = mat[0].size();
        while (1)
        {
                 if (j >= m) i += 2, j --, s = -s;
            else if (i >= n) i --, j += 2, s = -s;
            else if (i < 0) i ++, s = -s;
            else if (j < 0) j ++, s = -s;
            ans.emplace_back(mat[i][j]);
            if (i == n-1 && j == m-1) break;
            i -= s, j += s;
        }
        return ans;
    }
};